packages/proteus/src/util/RandomUtil.ts

```diff
@@ -28,3 +30,151 @@ export const random_bytes = (length: number): Uint8Array => {
   return new Uint8Array(crypto.randomBytes(length));
 };
+
+export class HmacDrbgSha256 {
+  private K: Uint8Array;
+  private V: Uint8Array;
+  private reseed_counter: number;
+
+  public constructor() {
+    this.K = new Uint8Array(32);
+    this.V = new Uint8Array(32);
+    this.reseed_counter = 0;
+  }
+
+  public instantiate(entropy?: Uint8Array, nonce?: Uint8Array, additional_input?: Uint8Array) {
+    this.K.fill(0);
+    this.V.fill(1);
+    this.update(new Uint8Array([...(entropy ?? []), ...(nonce ?? []), ...(additional_input ?? [])]));
+    this.reseed_counter = 1;
+  }
+
+  public reseed(entropy?: Uint8Array, additional_input?: Uint8Array) {
+    this.update(new Uint8Array([...(entropy ?? []), ...(additional_input ?? [])]));
+    this.reseed_counter = 1;
+  }
+
+  private update(additional_input?: Uint8Array) {
+    if (additional_input && additional_input.length > 34_359_738_368) {
+      throw new Error('Additional input too long.');
+    }
+    let K = sodium.crypto_auth_hmacsha256(new Uint8Array([...this.V, 0, ...(additional_input ?? [])]), this.K);
+    let V = sodium.crypto_auth_hmacsha256(this.V, K);
+    if (additional_input != null) {
+      K = sodium.crypto_auth_hmacsha256(new Uint8Array([...V, 1, ...(additional_input ?? [])]), K);
+      V = sodium.crypto_auth_hmacsha256(V, K);
+    }
+    this.K = K;
+    this.V = V;
+  }
+
+  public generate(length: number, additional_input?: Uint8Array) {
+    if (length > 1 << 16) {
+      throw new Error('Requested output length too long.');
+    }
+    if (this.reseed_counter > 281_474_976_710_656) {
+      throw new Error('Need reseeding.');
+    }
+    if (this.reseed_counter === 0) {
+      throw new Error('Need instantiating.');
+    }
+    if (additional_input) {
+      this.update(additional_input);
+    }
+    let temp = new Uint8Array();
+    while (temp.length < length) {
+      this.V = sodium.crypto_auth_hmacsha256(this.V, this.K);
+      temp = new Uint8Array([...temp, ...this.V]);
+    }
+    const retval = temp.slice(0, length);
+    this.update(additional_input);
+    this.reseed_counter++;
+    return retval;
+  }
+}
+
+/**
+ * A cryptographically secure pseudo-random number generator based DRBG_HMAC with SHA-2 256 bit.
+ * The construction meets the requirements of class DRG.3 (Killmann, Schindler, 2011), or
+ * class DRG.4, if it's regularly seeded from a PTG.2 or better.
+ *
+ * **Important note on usage**: This class is designed to be used as a singleton, instances should
+ * be retrieved from `get_instance()`. This ensures that the instance receives all external
+ * seeding.
+ */
+export class CSPRNG {
+  private static instance: CSPRNG;
+
+  /**
+   * Get the singleton instance of this class.
+   */
+  public static get_instance(): CSPRNG {
+    if (!CSPRNG.instance) {
+      CSPRNG.instance = new CSPRNG();
+    }
+
+    return CSPRNG.instance;
+  }
+
+  private readonly drbg: HmacDrbgSha256;
+  private seeded: Boolean;
+
+  public constructor() {
```

Comment:
Let's make the constructor `private` so that we are absolutely sure it's used as a singleton. 

Other abstractions might be more suitable for by-design singleton. 
Extracting to a separate file, have the instance created at the top level of the file and only exposing the public API would be one of them (I let you decide which solution you prefer between private constructor or external module)